### ISSUE TITLE
As a developer, I want a vendor staging site and I want to know how to deploy to it

### ISSUE THREAD

**Issue by alexsoble** [raft-tech/TANF-app#521]
## Tasks (in roughy chronological order)

1. [x] Write out Deployment plan for staging environment; get input from the dev team.
2. [x] Use the existing `deploy-cloudgov-docker.sh` script to deploy a staging site.
3. [x] Use the opportunity to improve deploy-related documentation along the way. 
4. [x] Once staging site launched using script, review ENV variable setup and consult with Tech Lead on any questions.
5. [x] Make tweaks as needed to staging site until the latest code is up and running properly in Cloud.gov
6. [x] Update CI/CD pipeline to deploy code automatically to staging
7. [x] Document new continuous deployment flow in How We Work and README, share out final continuous deployment flow to dev team 
8. [x] Seek Lauren’s PO approval and :ship:

## ACs 

+ [x] As a project team member who reviews dev work, I can visit the frontend of the staging site (https://tdp-frontend-vendor-staging.app.cloud.gov/) and log in
+ [x] As a project team member who reviews dev work, I can visit the backend of the staging site and see a sample API response 
+ [x] As a developer, I can easily find documentation about where the vendor staging site is
+ [x] As a developer, I can easily find documentation about when to deploy new code to the vendor staging site
+ [x] As a developer, I can easily find documentation about how new code is deployed to the vendor staging site
+ [x] As a developer, I can deploy new code to the vendor staging site via CircleCI

**Comment by alexsoble** [comment]
Proposed an ADR to document the overall deployment flow, was accepted and merged by dev team and PO today: https://github.com/HHS/TANF-app/pull/159

**Comment by alexsoble** [comment]
Tracking notes and adding additional code comments to the existing deploy script in this PR: https://github.com/HHS/TANF-app/pull/153

**Comment by alexsoble** [comment]
First pass at running the deploy script did deploy two applications to Cloud.gov, but the backend app state is Crashed:

```
tdp-frontend-vendor-staging - Deployed - Online
tdp-backend-vendor-staging - Deployed - Crashed
```

Error stack trace from the backend deploy:

```
FAILED
./deploy-cloudgov-docker.sh: line 100: generate_jwt_cert: command not found
```

**Comment by alexsoble** [comment]
Note -- split off and make a new ticket -- "as a dev standing up a new test/staging environment, I want a list of all the ENV variables I'll need and how to get them"

**Comment by alexsoble** [comment]
Created a new ticket — 609 — to address the current blocker. 

The deploy script for this project (`./deploy-cloudgov-docker.sh`) deploys a frontend and backend app. To run, the backend app requires S3 bucket configuration.  

In its current state, a dev can use the deploy script to automatically deploy a frontend and backend app, but the backend will remain in a crashed state unless it is supplied with credentials for an associated Cloud.gov-brokered S3 bucket.

**Comment by alexsoble** [comment]
Paired with @carltonsmith today on getting the S3 config set on the staging site. Unfortunately, even after S3 buckets were created and the necessary config was added, the backend site still crashed on startup -- this time, it needs database keys meaning a new provisioned database. 

@carltonsmith and I looked into the deploy script and saw that it is currently only configured for a single database per Space, whereas our Deploy Flow has separate deploy environments within the dev Space for ease of access management. 

We decided that the deploy script should be updated to allow it to handle multiple databases per Space. The deploy script will need an update anyway to add S3 provisioning as part of the deploy script. 

I'll be out of office all next week and won't be able to work on this, so I'll pass my work so far over to @carltonsmith and @hursey013 to prioritize with the rest of the team next week and carry forward in whatever way seems best!

**Comment by alexsoble** [comment]
Passing this work over to @hursey013 and @carltonsmith to drive forward as they see fit next week while I'm out! @carltonsmith mentioned that this felt like a high priority to him. Having a vendor staging site would help accelerate development by reducing the "crowding" problem (see ADR 008, Deployment Flow for more). 

Here is the current state of the work:

* Frontend and backend vendor staging apps deployed to Cloud.gov, but backend app is in a crashed state
* Backend app requires a database to be provisioned and db credentials added to be (hopefully!) uncrashed
* Three Draft PRs to:
  * Debug, clarify, and improve the deploy script (https://github.com/HHS/TANF-app/pull/153)
  * Set up automated deploys to the vendor staging sites once they're working (https://github.com/HHS/TANF-app/pull/170/)
  * Update documentation to reflect the new sites once all is done (https://github.com/HHS/TANF-app/pull/171/)

I got the chance to sync up with @carltonsmith and @hursey013 on all of these PRs and the overall state of the work earlier today. Appreciate being able to pass the baton forward on this work!

**Comment by alexsoble** [comment]
The documentation on env variable order of prioritization added in 510/backend is great and helps unblock this work!: https://github.com/HHS/TANF-app/pull/167/ :)

**Comment by alexsoble** [comment]
I manually created a Postgres DB for the backend vendor staging site and bound it to the app on cloud.gov. The backend vendor staging site is now un-crashed! :) 

```
name:              tdp-backend-vendor-staging
requested state:   started
routes:            tdp-backend-vendor-staging.app.cloud.gov
last uploaded:     Thu 11 Feb 10:23:29 CST 2021
stack:             
docker image:      lfrohlich/tdp-backend:raft-tdp-main

type:           web
instances:      1/1
memory usage:   256M
     state     since                  cpu    memory           disk           details
#0   running   2021-02-11T16:24:08Z   0.3%   132.4M of 256M   619.3M of 2G   
```

However, there are still at least two issues to debug:

+ When you log in at [tdp-frontend-vendor-staging.app.cloud.gov](https://tdp-frontend-vendor-staging.app.cloud.gov) and then log out, you are redirected to the wrong site -- [tdp-frontend.app.cloud.gov](https://tdp-frontend.app.cloud.gov) instead of [tdp-frontend-vendor-staging.app.cloud.gov](https://tdp-frontend-vendor-staging.app.cloud.gov). My hypothesis is that this is due to the Login.gov integration, and that we'll need to work on the Login.gov integration to handle multiple sites such as dev, staging, etc.
+ On the frontend, the Dockerfile hard-codes in [https://tdp-backend.app.cloud.gov/v1](https://tdp-backend.app.cloud.gov/v1) -- that will have to be extracted out and made more flexible. 

Once those two issues with the vendor staging site are addressed, and assuming no more crop up, we'll want to update CI/CD pipeline to handle automatic code deploy to staging and finish up this work!

**Comment by carltonsmith** [comment]
@alexsoble do you have a branch checked out for any work here or should I create a new one. That var should be extracted out to an env file for the environment on the frontend. I think there is a redirect env var on the backend we'll need to look at as well.

**Comment by alexsoble** [comment]
@carltonsmith For the frontend var, see: https://github.com/HHS/TANF-app/pull/184.

**Comment by alexsoble** [comment]
The vendor staging site was used to demo new work this week, and the dev team confirmed that they're good with the dev-related ACs... closing this one out! 

Thank you @jtwillis92 @carltonsmith and everyone else who pushed this past the finish line! 🎉
